jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java

```diff
@@ -651,105 +684,95 @@ public String toString()
-                                    if (last == 0 || (now - last) > TimeUnit.MILLISECONDS.toNanos(_idleTimeout))
-                                    {
-                                        if (_lastShrink.compareAndSet(last, now) && _threadsStarted.compareAndSet(size, size - 1))
-                                        {
-                                            shrink = true;
-                                            break loop;
-                                        }
-                                    }
+                                    if (_lastShrink.compareAndSet(last, now))
+                                        break loop;
                                 }
-                                job = idleJobPoll();
                             }
+
+                            job = _jobs.poll(_idleTimeout, TimeUnit.MILLISECONDS);
                         }
                     }
-                    finally
+
+                    // run job
+                    if (job != null)
                     {
-                        if (_threadsIdle.decrementAndGet() == 0)
+                        if (idle)
                         {
-                            startThreads(1);
+                            idle = false;
+                            if (_threadsIdle.decrementAndGet() == 0)
+                                startThreads(1);
                         }
+
+                        if (LOG.isDebugEnabled())
+                            LOG.debug("run {}", job);
+                        runJob(job);
+                        if (LOG.isDebugEnabled())
+                            LOG.debug("ran {}", job);
+
+                        // Clear interrupted status
+                        Thread.interrupted();
                     }
+
+                    if (!isRunning())
+                        break loop;
+
+                    job = _jobs.poll();
                 }
             }
             catch (InterruptedException e)
             {
-                ignore = true;
                 LOG.ignore(e);
             }
             catch (Throwable e)
             {
+                LOG.warn("Unexpected thread death: {} in {}", this, QueuedThreadPool.this);
                 LOG.warn(e);
             }
             finally
             {
-                if (!shrink && isRunning())
-                {
-                    if (!ignore)
-                        LOG.warn("Unexpected thread death: {} in {}", this, QueuedThreadPool.this);
-                    // This is an unexpected thread death!
-                    if (_threadsStarted.decrementAndGet() < getMaxThreads())
-                        startThreads(1);
-                }
+                if (idle)
+                    _threadsIdle.decrementAndGet();
+
                 removeThread(Thread.currentThread());
+
+                if (_threadsStarted.decrementAndGet() < getMinThreads())
+                    startThreads(1);
```

Comment:
When we are stopping, we don't want to create new threads, so `startThreads(1)` must be executed _only_ if we are running.
Can you please write a test case that verifies this? I.e. start the pool, start some threads, stop the pool and verify that no threads are running.

I prefer that there is only 1 check for `isRunning()`, so I think you should use a boolean variable to store the result of `isRunning()` and test it here to decide whether or not start a new thread.

Reply:
`startThreads()` already has an `isRunning()` check in it, so if the loop exits from the `isRunning()` check failing then no new threads will be started